### PR TITLE
New doc

### PR DESCRIPTION
**Pull request by vietj**
Fix inconsistencies throughout API + new documentation start

### CODE REVIEW

**Comment by vietj** [comment]
Please someone review

File: src/main/asciidoc/java/index.adoc [hunk 1]
```diff
@@ -0,0 +1,741 @@
+= Vert.x Codegen
+
+Vert.x Codegen is an annotation processing tool for processing Vert.x API and create API in different JVM lauguages.
+
+Vert.x polyglot langs use code generation for creating https://en.wikipedia.org/wiki/Shim_(computing)[shim] for APIs,
+that is thin wrappers in front of the API type.
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: are

File: src/main/asciidoc/java/index.adoc [hunk 2]
```diff
@@ -0,0 +1,741 @@
...
+
+A shim is the adaptation of a Vert.x API in a JVM language. A shim uses code generation to generate
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: I would say 'to' and not 'in'

File: src/main/asciidoc/java/index.adoc [hunk 3]
```diff
@@ -0,0 +1,741 @@
...
+the shim in its language from a Vert.x API, code generation is based on Java annotation processing.
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: code / API instead of the second 'shim'

File: src/main/asciidoc/java/index.adoc [hunk 4]
```diff
@@ -0,0 +1,741 @@
...
+[source,java]
+----
+ModuleGen(name = "acme", groupPackage="com.acme")
+package com.acme.myservice;
+
+import io.vertx.codegen.annotations.ModuleGen;
+----
+
+The module _name_ is a namespace for shims that don't support package like naming, like JavaScript `acme-js`
+or Ruby `acme`, whereas the _group package_ is determines the created package, for instance
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: remove the 'is' from 'is determines'

File: src/main/asciidoc/java/index.adoc [hunk 5]
```diff
@@ -0,0 +1,741 @@
...
+for Groovy `com.acme.groovy.myservice`. The group package must be a prefix of the annotated module.
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: such as `com.acme.groovy.myservice` in Groovy. (avoid the 2 for)

File: src/main/asciidoc/java/index.adoc [hunk 6]
```diff
@@ -0,0 +1,741 @@
...
+
+Vert.x modules uses the reserved name _vertx-XYZ_ and the reserved group package _io.vertx_.
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: use

File: src/main/asciidoc/java/index.adoc [hunk 7]
```diff
@@ -0,0 +1,741 @@
...
+
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: it's unclear what is a vert.x modules. vert.x official modules ? Modules form the official vert.x stack ?

File: src/main/asciidoc/java/index.adoc [hunk 8]
```diff
@@ -0,0 +1,741 @@
...
+An API module contains various Vert.x types annotated with `link:../../apidocs/io/vertx/codegen/annotations/VertxGen.html[@VertxGen]`
+or `link:../../apidocs/io/vertx/codegen/annotations/DataObject.html[@DataObject]`.
+
+NOTE: using Maven coordinates for name and group package is encouraged: the name corresponding to the
+Maven _artifactId_ and the group package corresponding to the `groupId`.
+
+=== Data objects
+
+_Data objects_ are Java class whose only purpose are to be data containers of various value types. A data object
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: A _Data object_ is a Java class with the only purpose to be a container for data. Date objects are transformed to and from JSON.

File: src/main/asciidoc/java/index.adoc [hunk 9]
```diff
@@ -0,0 +1,741 @@
...
+is mapped in some languages to Json.
+
+In its simplest form, a data object is a Java class following these rules:
+
+1. the class is annotated with `link:../../apidocs/io/vertx/codegen/annotations/DataObject.html[@DataObject]`
+2. provide a zero argument constructor
+3. provide a constructor with the `io.vertx.core.json.JsonObject` argument
+4. provide a copy constructor with the exact same type
+
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: don't you need also `toJson` ?

File: src/main/asciidoc/java/index.adoc [hunk 10]
```diff
@@ -0,0 +1,741 @@
...
+A data object can be also be an interface annotated with `@DataObject`, this is useful when multiple inheritance
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: can also be an

File: src/main/asciidoc/java/index.adoc [hunk 11]
```diff
@@ -0,0 +1,741 @@
...
+is needed. For instance Vert.x Core defines the `KeyCertOptions` and `TrustOptions` data object interfaces and the
+`JksOptions` is a data object class that implements both of them because a Java keystore provides support for both.
+
+Data object can also inherit other data objects, when it does so it inherits naturally its properties.
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: inherit from other data objects. It inherits from the properties of the parent classes.

File: src/main/asciidoc/java/index.adoc [hunk 12]
```diff
@@ -0,0 +1,741 @@
...
+Here is the list of supported property single valued types:
+
+1. any primitive or boxed primitive type
+2. `java.lang.String`
+3. `io.vertx.core.json.JsonObject` and `io.vertx.core.json.JsonArray`
+4. the specific `io.vertx.core.buffer.Buffer` type providing support for byte array
+5. Java enums
+6. another data object
+
+In addition a data object can also have multi valued properties as a `java.util.List<V>` or a
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: multi-valued

File: src/main/asciidoc/java/index.adoc [hunk 13]
```diff
@@ -0,0 +1,741 @@
...
+`java.util.Map<String, V>` where the `<V>` is a supported single valued type or `java.lang.Object`
+that stands for anything converted by `io.vertx.core.json.JsonObject` and `io.vertx.core.json.JsonArray`.
+
+List multi valued properties can be declared via a _setter_ :
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: multi-valued

File: src/main/asciidoc/java/index.adoc [hunk 14]
```diff
@@ -0,0 +1,741 @@
...
+.a multi valued adder
+[source,java]
+----
+public WebServerOptions addCertificate(String certificate) {
+  this.certificates.add(certificate);
+  return this;
+}
+----
+
+Map multi valued properties can be declared via a _setter_.
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: Map properties can only be declared with a _setter_.

File: src/main/asciidoc/java/index.adoc [hunk 15]
```diff
@@ -0,0 +1,741 @@
...
+
+NOTE: these examples uses a _fluent_ return types for providing a better API, this is not mandatory but
+encouraged
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: .

File: src/main/asciidoc/java/index.adoc [hunk 16]
```diff
@@ -0,0 +1,741 @@
...
+generation happens for the data object properties, not for the ancestor properties, unless `inheritConverter`
+is set: `@DataObject(generateConverter=true,inheritConverter=true)`.
+
+The converter is named by appending the `Converter` suffix to the data object class name, e.g,
+`ContactDetails` -> `ContactDetailsConverter`. The generated converter has two static methods:
+
+- `public static void fromJson(JsonObject json, ContactDetails obj)`
+- `public static void toJson(ContactDetails obj, JsonObject json)`
+
+The former should be used in the json constructor, the later in the `toJson` method.
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: Maybe you should add an example with a constructor taking the JSON object are argument and the toJson method.

File: src/main/asciidoc/java/index.adoc [hunk 17]
```diff
@@ -0,0 +1,741 @@
...
+
+An API type is a Java interface annotated with `link:../../apidocs/io/vertx/codegen/annotations/VertxGen.html[@VertxGen]`.
+
+Vert.x provides a async / non blocking / polyglot programming model, code generated API shall follow some
+rules to make this possible:
+
+1. the API must be described as a set of Java interfaces, classes are not permitted
+2. nested interfaces are not permitted
+3. All interfaces to have generation performed on them must be annotated with the `io.vertx.codegen.annotations.VertxGen` annotation
+4. fluent methods (methods which return a reference to this) must be annotated with the `io.vertx.codegen.annotations.Fluent` annotation
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: `this`

File: src/main/asciidoc/java/index.adoc [hunk 18]
```diff
@@ -0,0 +1,741 @@
...
+5. methods where the return value must be cached in the API shim must be annotated with the `io.vertx.codegen.annotations.CacheReturn` annotation
+6. only certain types are allowed as parameter or return value types for any API methods
+7. custom enums should be annotated with `@VertxGen`, although this is not mandatory to allow the usage of existing Java enums
+
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: nested enums are not supported.
default implementation are supported.

File: src/main/asciidoc/java/index.adoc [hunk 19]
```diff
@@ -0,0 +1,741 @@
...
+An API type can be generic or declare generic methods, type parameters must be unbounded, e.g
+`<N extends Number>` is forbidden.
+
+In the perspective of codegen, Java types can be categories as follow:
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: categorized

File: src/main/asciidoc/java/index.adoc [hunk 20]
```diff
@@ -0,0 +1,741 @@
...
+
+. _basic_ type : any primitive/boxed type or `java.lang.String`
+. _json_ type : `io.vertx.core.json.JsonObject` or `io.vertx.core.json.JsonArray`
+. _api_ type : any type annotated with `io.vertx.codegen.annotations.VertxGen`
+. _data object_ type : any type annotated with `io.vertx.codegen.annotations.DataObject`
+. _enum_ type : any Java enum
+. _collection_ type : `java.util.List<V>`, `java.util.Set<V>` or `java.util.Map<String, V>`
+
+Parameterized types are supported but wildcards are not, that is the following type arguments declarations
+are forbidden:
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: **forbidden**

File: src/main/asciidoc/java/index.adoc [hunk 21]
```diff
@@ -0,0 +1,741 @@
...
+Parameterized types are only supported for _api_ generic types and _collection_ types.
+
+Type variables are allowed and carry a special meaning: a type variable is a dynamic form of a _basic_ type and
+_json_ type.
+
+==== Inheritance
+
+_api_ type can extend other _api_ types.
+
+An _api_ type can either be *concrete* or *abstract*, such information is importan for languages not
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: important

File: src/main/asciidoc/java/index.adoc [hunk 22]
```diff
@@ -0,0 +1,741 @@
...
+The following return types are allowed:
+
+. `void` type
+. any _basic_ type
+. any _api_ type or parameterized _api_ type having type variable parameters
+. any _json_ type
+. the `java.lang.Throwable` type
+. any _enum_ type
+. any _data object_ type
+. an unbounded type variable
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: to explain

File: src/main/asciidoc/java/index.adoc [hunk 23]
```diff
@@ -0,0 +1,741 @@
...
+. a `java.util.List<V>`, `java.util.Set<V>` or `java.util.Map<String, V>` where `<V>` can be a _basic_ type,
+a _json_ type, an _API_ type, an _enum_ type or a _data object_ type
+
+==== Method overloading
+
+Some languages don't support method overloading at all. Ruby, JavaScript or  Ceylon to name a few of them.
+However the same restrication for Vert.x API would limit API usability.
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: restriction

File: src/main/asciidoc/java/index.adoc [hunk 24]
```diff
@@ -0,0 +1,741 @@
...
+Here is an example of possible ambiguity:
+
+.an overload check failure
+[source,java]
+----
+void add(int x, int y);
+void add(double x, double y);
+----
+
+The JavaScript language use the type number in obth case: at runtime there is no possibility for the
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: both

File: src/main/asciidoc/java/index.adoc [hunk 25]
```diff
@@ -0,0 +1,741 @@
...
+
+=== Ignored methods
+
+Methods annotated with `link:../../apidocs/io/vertx/codegen/annotations/GenIgnore.html[@GenIgnore]` are simply ignored by codegen, this
+is useful when the API provides Java specific methods, for instance a method uses a type not permitted
+by codegen.
+
+== Shim proxies
+
+A code generated API create shim proxies that delegates method invocation to the proxied API.
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: creates shim proxies delegating method invocation to the proxies API.

File: src/main/asciidoc/java/index.adoc [hunk 26]
```diff
@@ -0,0 +1,741 @@
...
+
+A JavaScript generated shim could look like:
+
+.the JavaScript shim
+[source,javascript]
+----
+var JBuffer = io.vertx.core.buffer.Buffer;
+var Buffer = function(j_val) {
+
+  var j_buffer = j_val;
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: // delegate object

File: src/main/asciidoc/java/index.adoc [hunk 27]
```diff
@@ -0,0 +1,741 @@
...
+}
+
+module.exports = Buffer;
+----
+
+The static `buffer` method is translated into the `buffer` method of the `Buffer` module, this method
+delegates the call to the Java static method and returns a `Buffer` proxy wrapping the returned buffer.
+
+The instance `length` method is translated into the `length` method of the proxy instance, this method
+delgates the call to the Java instance method of the proxied buffer and simply returns the value. The
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: delegates

File: src/main/asciidoc/java/index.adoc [hunk 28]
```diff
@@ -0,0 +1,741 @@
...
+The codegen processor _validates_ annotated Java program elements (i.e type declaration) and _transforms_  them into models:
+
+1. `ClassModel`
+2. `DataObjectModel`
+3. `EnumModel`
+4. `PackageModel`
+5. `ModuleModel`
+6. `ProxyModel`
+
+Models are processed by _MVEL_ templates, when a template is executed it gets access to implicit properties
```
Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: add a link to the MVEL doc (now that there is one)

**Comment by cescoffier** [comment]
I've made a few comments. Looking forward seeing the next step of the documentation.

File: src/main/asciidoc/java/index.adoc (hunk 22 above)
Comment (vietj) [inline src/main/asciidoc/java/index.adoc]: what should I explain :) ?

Comment (cescoffier) [inline src/main/asciidoc/java/index.adoc]: What you call "an unbounded type variable". Unbounded has many semantic.

Comment (vietj) [inline src/main/asciidoc/java/index.adoc]: ah ok, unbounded in the context of type variable and it should rather be "unbounded type parameter" https://docs.oracle.com/javase/tutorial/java/generics/bounded.html . will correct and add link

**Pull request merged**